pkg/controllers/cluster/actions/upgrade_version.go

```diff
@@ -3,54 +3,966 @@ package actions
 import (
 	"context"
 	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+	"time"
 
+	"github.com/blang/semver"
+	"github.com/gocql/gocql"
 	"github.com/pkg/errors"
 	"github.com/scylladb/go-log"
+	"github.com/scylladb/go-set/strset"
+	"github.com/scylladb/gocqlx/v2"
 	scyllav1alpha1 "github.com/scylladb/scylla-operator/pkg/api/v1alpha1"
 	"github.com/scylladb/scylla-operator/pkg/controllers/cluster/resource"
 	"github.com/scylladb/scylla-operator/pkg/controllers/cluster/util"
 	"github.com/scylladb/scylla-operator/pkg/naming"
+	"github.com/scylladb/scylla-operator/pkg/scyllaclient"
+	"github.com/scylladb/scylla-operator/pkg/util/fsm"
+	"github.com/scylladb/scylla-operator/pkg/util/parallel"
+	"github.com/scylladb/scylla-operator/pkg/util/timeutc"
 	appsv1 "k8s.io/api/apps/v1"
 	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 const (
 	ClusterVersionUpgradeAction = "rack-version-upgrade"
+	actionTimeout               = time.Minute
 )
 
 type ClusterVersionUpgrade struct {
-	Cluster *scyllav1alpha1.ScyllaCluster
-	logger  log.Logger
+	Cluster        *scyllav1alpha1.ScyllaCluster
+	ScyllaClient   *scyllaclient.Client
+	ClusterSession cqlSession
+
+	ipMapping map[string]string
+
+	currentRack *scyllav1alpha1.RackSpec
+	currentNode *corev1.Pod
+
+	cc         client.Client
```

Comment:
I'd consider moving Client to scylla package

Reply:
this is k8s apiserver client